[Backend]: Change all PUT routes to POST

michellemtchai opened the pull request:
Please see issue #182 for more details.

michellemtchai committed:
fix(api): fix issues in controller
fix applications & instructors controller

michellemtchai committed:
feat(api): set up post delete route
setup delete routes for the following controllers:
applicants
application
assignment
instructor
position_preference
postition_template
position
reporting_tag
sessions
wage_chunk

michellemtchai committed:
feat(api): make :update into post routes
make update into post routes for the following controllers:
applicants
applications
assignment
instructor
position_preference
position_templates
reporting_tags
wage_chunks

michellemtchai committed:
fix(api): delete related tables for positions & applications
setup delete for tables associated with positions and applications in their delete routes

michellemtchai committed:
fix(api): add instructor_ids as params for update/create position
allow linking with instructors when creating/updating positions

michellemtchai committed:
feat(api): add get all for each controller
create a general get all :index for the controllers

michellemtchai committed:
feat(api): render related tables attributes for position & application
add attributes from position_data_for_matching and position_data_for_ad
add attributes from applicant_data_for_matching

siefkenj commented on api/app/controllers/api/v1/instructors_controller.rb:
```diff
@@ -54,10 +68,18 @@ def instructors_by_position
+                render_success(result)
+            else
+                instructor.destroy!
+                render_error(instructor.errors, error)
```
Why not `render_error(instructor.errors.full_messages.join("; "))`?

michellemtchai replied:
Good catch. I'll fix that now. Thanks.

michellemtchai committed:
fix(api): fix error message for instructor_create
this give a string message instead of a json message

George334456 commented on api/app/controllers/api/v1/applicants_controller.rb:
```diff
@@ -16,6 +16,11 @@ def index
+
             applicant = Applicant.new(applicant_params)
             if applicant.save # passes Applicant model validation
                 render_success(applicant)
```
For line 34 in this file, couldn't we techinically hit PUT/PATCH on applicants/:id?

michellemtchai replied:
Yes, we could do that. I guess I'll just remove that if statement.

George334456 commented on api/app/controllers/api/v1/applicants_controller.rb:
```diff
@@ -16,6 +16,11 @@ def index
 
         # POST /applicants
         def create
+            # if we passed in an id that exists, we want to update
```
To clarify this defines two routes:
`/applicants` which will create an applicant based on the things passed in.
`/applicants/:id` which will update an existing applicant based on ID.

Side note: if you pass in an `id` that does not exist, this will behave like `/applicants` and ignore the id parameter. Is this intended or should we give a `id not found` error?

michellemtchai replied:
I don't know if the former or latter is the intended behaviour. Usually, I would think it is the latter, which will make this implementation. However, when I was writing this section, @siefkenj said I should follow the format he used for this `update` vs. `create` separation. You would have to ask him for more details.

michellemtchai replied:
I ended up creating `/applicants/:id` sort of `update` routes instead of rerouting via `create`.

George334456 commented on api/app/controllers/api/v1/applicants_controller.rb:
```diff
@@ -34,6 +39,17 @@ def update
             end
         end
+
+        # /applicants/delete or /applicants/:id/delete
```
`/applicants/delete` can't happen right? Because you require the id.

michellemtchai replied:
Actually, it can thanks to the `config/routes.rb`. Though, I have no idea why there are two ways to delete. @siefkenj was the one who set up this structure. You would have to ask him if you want to know why it was setup this way.

michellemtchai replied:
I think I'll delete the former. It's kind of redundant anyways.

michellemtchai replied:
Nevermind, I just realized the `/applicants/delete` exist because without it, this route is taken as `/applicants/:id`, where `id=delete`...

michellemtchai replied:
I deleted the `/applicants/delete` sort of route and kept only the `/applicants/:id/delete` type of routes.

George334456 commented on api/app/controllers/api/v1/applicants_controller.rb:
```diff
@@ -34,6 +39,17 @@ def update
+        def delete
+            params.require(:id)
+            applicant = Applicant.find(params[:id])
+            if session.destroy!
```
Why destroy the `session`? Aren't we destroying the `applicant`?

michellemtchai replied:
Good catch. This was a mistake. I'll fix that. Thanks.

michellemtchai replied:
Fixed.

George334456 commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -84,11 +107,20 @@ def matching_data_update_params
+        def all_applications
+            exclusion = [:id, :created_at, :updated_at, :application_id]
+            return Application.order(:id).map do |entry|
+                matching = ApplicantDataForMatching.find_by(application_id: entry[:id])
```
`Application` has one `ApplicantDataForMatching`.

Couldn't we do `matching = entry.ApplicantDataForMatching`?

michellemtchai replied:
I'll fix this. Thanks for catching this.

michellemtchai replied:
Done.

George334456 commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -16,7 +16,14 @@ def index
         # POST /applications
         def create
+            # if we passed in an id that exists, we want to update
+            if params[:id] && Application.exists?(params[:id])
```
Same comment for `POST applicant`

George334456 commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -16,7 +16,14 @@ def index
+
+            params.require(:applicant_id)
             if invalid_id(Session, :session_id) then return end
+            if invalid_id(Applicant, :applicant_id) then return end
```
Is `invalid_id` defined somewhere?

michellemtchai replied:
It is in `app/controllers/concerns/response.rb`.

George334456 commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -16,7 +16,14 @@ def index
+            end
+
+            params.require(:applicant_id)
             if invalid_id(Session, :session_id) then return end
```
I'm assuming that something like `:session_id` is passed in the JSON body right? If not, where did we store it?

michellemtchai replied:
Yes, the `session_id` is stored in the JSON body.

George334456 commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -36,7 +43,7 @@ def create
         def update
             application = Application.find(params[:id])
-            matching = ApplicantDataForMatching.find(application[:id])
+            matching = ApplicantDataForMatching.find_by!(application_id: application[:id])
```
Again, `Application` has a `has one` relationship to `ApplicantDataForMatching`. Couldn't we just do:
`application.ApplicantDataForMatching`?

michellemtchai replied:
Good point. Thanks for catching that. I'll change it.

michellemtchai replied:
Done.

George334456 commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -47,6 +54,22 @@ def update
             end
         end
 
+        # /applications/delete or /applications/:id/delete
```
Again, if `id` is required, we should be never able to access `applications/delete` right?

George334456 commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -47,6 +54,22 @@ def update
+            params.require(:id)
+            application = Application.find(params[:id])
+            if application
+                applicant_data_for_matching = ApplicantDataForMatching.find_by(
```
Another place to potentially change to `application.applicant_data_for_matching`

michellemtchai replied:
I'll fix this as well. Thanks.

michellemtchai replied:
Done.

George334456 commented on api/app/controllers/api/v1/applications_controller.rb:
```diff
@@ -47,6 +54,22 @@ def update
+            if application
+                applicant_data_for_matching = ApplicantDataForMatching.find_by(
+                    application_id: params[:id])
+                applicant_data_for_matching.destroy!
```
Also are we guaranteed applicant_data_for_matching? If there isn't one, won't `destroy` fail and muck up everything down stream? Should we gate this with an `applicant_data_for_matching.destroy! if applicant_data_for_matching`?

michellemtchai replied:
It should be guaranteed because the controller is assuming that our tables are always created via the API routes. If there was an error in the `applicant_data_for_matching` creation, it should have shown in the error message and the corresponding `application` should have been deleted.

George334456 replied:
I meant it in the sense that couldn't we have an application without an applicant data for matching?

michellemtchai replied:
Not really. @siefkenj's explanation for this was that the `application` and `applicant_data_for_matching` are really meant to be the same table and that they are only separated for semantics.

George334456 commented on api/app/controllers/api/v1/assignments_controller.rb:
```diff
@@ -16,6 +16,11 @@ def index
         # POST /assignments
         def create
+            # if we passed in an id that exists, we want to update
+            if params[:id] && Assignment.exists?(params[:id])
```
Again, if an invalid id is passed in, won't this just do a regular create?

michellemtchai replied:
Yes. I'm assuming that what we want(?) The initial structure for this setup was made by @siefkenj, so you have to ask him about the specifics on why it was designed this way.

George334456 replied:
Hmm, this will have to be tested and kept consistent on live. If this is the supposition, we need to make sure that this is confirmed.

michellemtchai replied:
I got rid of the `create` rerouting to `update` and made the routes directly call `POST /tables/:id` instead.

George334456 commented on api/app/controllers/api/v1/assignments_controller.rb:
```diff
@@ -44,6 +49,17 @@ def update
+        def delete
+            params.require(:id)
+            assignment = Assignment.find(params[:id])
+            if session.destroy!
```
Why are we destroying the session? Shouldn't we be destroying the assignment?

michellemtchai replied:
Good catch. I'll fix this. Thanks.

George334456 commented on api/app/controllers/api/v1/instructors_controller.rb:
```diff
@@ -14,30 +14,44 @@ def index
+                update and return
+            end
+
+            if not params.include?(:position_id)
```
To get this straight, if there isn't a position_id in the parameters, we would create a new instructor.

If there is a position_id in the parameters, then we are adding an instructor to that position.

Is that correct?

michellemtchai replied:
It's more like if there isn't a `position_id`, we are creating a new `instructor` and adding that new `instructor` into `position`.

George334456 replied:
If there is no position_id, how can you add an instructor to the position since you don't know the `position_id`?

michellemtchai replied:
Sorry. I just realized that I made a mistake in my reply. I was referring to when there **is** a `position_id`.

George334456 commented on api/app/controllers/api/v1/instructors_controller.rb:
```diff
@@ -14,30 +14,44 @@ def index
+        def delete
+            params.require(:id)
+            instructor = Instructor.find(params[:id])
+            if session.destroy!
```
Same as previous.

George334456 commented on api/app/controllers/api/v1/instructors_controller.rb:
```diff
@@ -14,30 +14,44 @@ def index
-                instructor.destroy!
-                render_error(instructor.errors, instructors_by_position)
+                position = Position.find(params[:position_id])
+                instructor = position.instructors.new(instructor_params)
```
Does this syntax create a new instructor from the instructor_params? If my previous comment was correct, then I thought you could only do:
```
instructor = Instructor.new(instructor_params)
instructor_create(instructor, true)
position.instructors.append(instructor)
position.save!
```
To create a new instructor, then add a new instructor to the position.

michellemtchai replied:
This syntax does create a new `instructor` from `instructor_params`. I'll check out if it works differently from the code you wrote, though.

George334456 replied:
I'm just wondering if the position with a new instructor is really being saved. i know you save the instructor, but I don't know if you save the fact that this position now has a new instrucotr.

michellemtchai replied:
I think the `save` is actually a misnomer. It's not really saving to the db, `save` actually just checks if the data matching the table constraints. `new` already inserted the data into the db.

George334456 replied:
https://guides.rubyonrails.org/active_record_basics.html#create

According to the docs though new will not insert the data to the database? Did you try an `object.reload` to reload it from database and check that the changes have propagated?

michellemtchai replied:
Okay. I'll check that.

michellemtchai replied:
I tried the `reload`, but it seems like `save` is only really relevant when creating a table. If you update, say, `position.instructors` after creation, it does persist.

George334456 commented on api/app/controllers/api/v1/position_preferences_controller.rb:
```diff
@@ -4,8 +4,18 @@ module Api::V1
+        # POST /add_preference and /position_preferences
         def create
+            # if we passed in an id that exists, we want to update
+            if params[:id] && PositionPreference.exists?(params[:id])
```
Same as previous

George334456 commented on api/app/controllers/api/v1/position_preferences_controller.rb:
```diff
@@ -28,6 +38,17 @@ def update
         end
 
+        # /position_preferences/delete or /position_preferences/:id/delete
+        def delete
```
Same as previous

George334456 commented on api/app/controllers/api/v1/position_templates_controller.rb:
```diff
@@ -6,11 +6,21 @@ class PositionTemplatesController < ApplicationController
+        # POST /add_position_template and /position_templates
         def create
+            # if we passed in an id that exists, we want to update
+            if params[:id] && PositionTemplate.exists?(params[:id])
```
Same as previous

George334456 commented on api/app/controllers/api/v1/position_templates_controller.rb:
```diff
@@ -38,6 +48,17 @@ def update
         end
 
+        # /position_templates/delete or /position_templates/:id/delete
+        def delete
```
Same as previous

George334456 commented on api/app/controllers/api/v1/position_templates_controller.rb:
```diff
@@ -38,6 +48,17 @@ def update
+        def delete
+            params.require(:id)
+            position_template = PostionTemplate.find(params[:id])
+            if session.destroy!
```
Same as previous.

George334456 commented on api/app/controllers/api/v1/positions_controller.rb:
```diff
@@ -24,6 +29,7 @@ def create
                 render_error(position.errors.full_messages.join("; "))
                 return
             end
+            update_instructors_ids(position)
```
Don't you have to save once you've updated the instructor ids for the position?

michellemtchai replied:
That is actually a very interesting point. I'll check if save affects anything.

michellemtchai replied:
While it doesn't seem like saving changes the persistence in the db here, should I `save` anyways?

George334456 replied:
I guess we couuuuuuuuuuuuuuuuuuuuuuuld. We'll have to test it somehow. I feel that something will pop up in prod when we try LUL

George334456 commented on api/app/controllers/api/v1/reporting_tags_controller.rb:
```diff
@@ -4,8 +4,18 @@ module Api::V1
+        # POST /add_reporting_tag and /reporting_tags
         def create
+            # if we passed in an id that exists, we want to update
+            if params[:id] && ReportingTag.exists?(params[:id])
```
Same as above.

George334456 commented on api/app/controllers/api/v1/reporting_tags_controller.rb:
```diff
@@ -29,6 +39,17 @@ def update
+        def delete
+            params.require(:id)
+            reporting_tag = ReportingTag.find(params[:id])
+            if session.destroy!
```
Same as above.

George334456 commented on api/app/controllers/api/v1/wage_chunks_controller.rb:
```diff
@@ -36,6 +41,17 @@ def update
+        def delete
+            params.require(:id)
+            wage_chunk = WageChunk.find(params[:id])
+            if session.destroy!
```
Same as above.

George334456 commented on api/app/controllers/api/v1/wage_chunks_controller.rb:
```diff
@@ -14,8 +14,13 @@ def index
-        # POST /add_wage_chunk
+        # POST /add_wage_chunk and /wage_chunks
         def create
+            # if we passed in an id that exists, we want to update
```
Same as above

George334456 commented on api/app/controllers/api/v1/sessions_controller.rb:
```diff
@@ -39,12 +39,13 @@ def update
 
         # /sessions/delete or /sessions/:id/delete
         def delete
```
Again, would `delete` work without an `id` in the url?

George334456 reviewed (changes requested):
Left some comments. Let's discuss them whenever you have time.

michellemtchai committed:
fix(api): fix typo & add update alternative route
fix typo in :delete for controllers
fix typo in controller comments for :update
create routes for /table/:id

michellemtchai committed:
fix(api): remove unnecessary create routes
remove create route not specified in API document, such as:
/reporting_tags
/wage_chunks

michellemtchai committed:
fix(api): use return if x for invalid_id checks
use one return if x to make invalid_id check more readable

michellemtchai committed:
fix(api): remove /tables/delete routes
keep only POST /tables/:id/delete for deleting entries in db

michellemtchai committed:
fix(api): use one-to-one link for position & application
use one-to-one relationship to minimize use of find_by
render full position/application data

michellemtchai committed:
fix(api): fix add_instruction given position
fix issue with position.instructors.new()

michellemtchai committed:
fix(api): fix typo in delete
fix typo in delete method in reporting_tag and wage_chunk controllers

michellemtchai committed:
fix(api): use better if condition for instructor_ids
use (params.include?(:instructor_ids)) vs (params[:instructor_ids])

michellemtchai committed:
fix(api): clean up position/application_data helpers
use with_indifferent_access to avoid multiple casting

michellemtchai committed:
fix(api): fix error in implementing valid_wage_chunk

michellemtchai committed:
fix(api): revert to rerouting & /<tables>/delete
removed POST /<tables>/:id for update in routes.rb
changed POST /<tables>/:id/delete to POST /<tables>/delete
add back require params[:id] in delete functions
add back if statements to reroute create to update

michellemtchai committed:
fix(api): add back plain create option for each controller
make POST /<table> for upsert available for every controller

michellemtchai committed:
fix(api): use proper params keys check for upsert
use params.has_keys?() because the original didn't catch the presence of :id
